tracer/src/Datadog.Tracer.Native/cor_profiler.cpp

```diff
@@ -2173,6 +2145,81 @@ HRESULT CorProfiler::RewriteForDistributedTracing(const ModuleMetadata& module_m
     return hr;
 }
 
+HRESULT CorProfiler::RewriteForTelemetry(const ModuleMetadata& module_metadata, ModuleID module_id)
+{
+    HRESULT hr = S_OK;
+
+    if (IsDebugEnabled())
+    {
+        LogManagedProfilerAssemblyDetails();
+    }
+
+    //
+    // *** Get Instrumentation TypeDef
+    //
+    mdTypeDef instrumentationTypeDef;
+    hr = module_metadata.metadata_import->FindTypeDefByName(instrumentation_type_name.c_str(),
+                                                            mdTokenNil, &instrumentationTypeDef);
+    if (FAILED(hr))
+    {
+        Logger::Warn("Error rewriting for Telemetry on getting Instrumentation TypeDef");
+        return hr;
+    }
+
+    //
+    // *** GetNativeTracerVersion MethodDef ***
+    //
+    constexpr COR_SIGNATURE getNativeTracerVersion[] = {IMAGE_CEE_CS_CALLCONV_DEFAULT, 0, ELEMENT_TYPE_STRING};
+    mdMethodDef getNativeTracerVersionMethodDef;
+    hr = module_metadata.metadata_import->FindMethod(instrumentationTypeDef, WStr("GetNativeTracerVersion"),
+                                                     getNativeTracerVersion, 3, &getNativeTracerVersionMethodDef);
+    if (FAILED(hr))
+    {
+        Logger::Warn("Error rewriting for Telemetry on getting GetNativeTracerVersion MethodDef");
+        return hr;
+    }
+
+    // Define NativeTracerVersion as a string
+    mdString nativeTracerVersionToken;
+    const auto nativeTracerVersion = ToWSTRING(PROFILER_VERSION);
+    hr = module_metadata.metadata_emit->DefineUserString(
+        nativeTracerVersion.c_str(), static_cast<ULONG>(nativeTracerVersion.length()), &nativeTracerVersionToken);
+
+    if (FAILED(hr))
+    {
+        Logger::Warn("Error rewriting for Telemetry on DefineUserString for NativeProfilerVersion");
+        return hr;
+    }
+
+    ILRewriter methodRewriter(this->info_, nullptr, module_id, getNativeTracerVersionMethodDef);
+    methodRewriter.InitializeTiny();
+
+    // Modify first instruction from ldstr "None" to ldstr PROFILER_VERSION
+    ILRewriterWrapper wrapper(&methodRewriter);
+    wrapper.SetILPosition(methodRewriter.GetILList()->m_pNext);
+    wrapper.LoadStr(nativeTracerVersionToken);
+    wrapper.Return();
```

Comment:
This will append but not replace the instruction, so you will keep the previous IL code but at the end of the method. Because the code already have a `ldstr` ("None") I think is better to check the first instruction opcode, and just modify the token to the new `nativeTracerVersionToken`. For that you don't need the `ILRewriterWrapper`

Reply:
Done in https://github.com/DataDog/dd-trace-dotnet/pull/3643/commits/ded77513426ebbd54ead747127bd0ae06289964b